lib/Sema/PreCheckExpr.cpp

```diff
@@ -700,6 +700,17 @@ Expr *TypeChecker::resolveDeclRefExpr(UnresolvedDeclRefExpr *UDRE,
   }
 
   if (AllMemberRefs) {
+    if (Name.isSimpleName(Context.Id_self) &&
+        UDRE->getFunctionRefKind() == FunctionRefKind::Unapplied) {
```

Comment:
`UDRE` is always either `Unapplied` if it's just a name, or `Compound` if it's a name with labels e.g. `test(a:)`. I think you need to figure out what the parent is here to be able to conclude whether it's called or not, so the check might be easier to implement in the callee of `resolveDeclRefExpr` - https://github.com/apple/swift/blob/main/lib/Sema/PreCheckExpr.cpp#L1139-L1140 if `self` has been unambiguously resolved to a DeclRefExpr that has a function type and it's not attached as `Fn` to a `CallExpr` or `UnresolvedMemberExpr` (maybe it's possible to have `... = .self(...)`)

Reply:
>  that has a function type
@xedin I'm trying to work out how to implement this check, if I do this:
`if (resolved->getType() && resolved->getType()->is<FunctionType>())`
It always seems to be false? Any advice would be much appreciated!

(Side note just in case this informs your recommendation: I've updated my PR so that it identifies whether the function is called on your advice, it seems the type of resolved we need to check for is `UnresolvedDotExpr` rather than `DeclRefExpr`)

Reply:
The check is always false because `UnresolvedDotExpr` doesn't have a concrete type resolved yet. That's why I mentioned that the only safe way to do this would be to use `DeclRefExpr` which is used when there is only one candidate for a particular reference so it could be resolved early.

To diagnose situations when there are multiple choices regarding to what `self` could be a given context, you'd have to move your check to `MiscDiagnostics` as a separate pass over a type-checker version of the AST, it would find all `DeclRefExpr` that refer to `self` and check that they are wrapped in a `CallExpr`. Does it make sense?

Reply:
thanks a bunch @xedin for the suggestions, I've implemented this and have marked the PR as ready to review as it seems to work / can't find any edge cases

Reply:
thanks a bunch for the suggestions @xedin, I've marked as ready for review because it seems to work and I can't find any false positives.

(I tried = .`.. = .self(...)` as suggested in this comment https://github.com/apple/swift/pull/37992#discussion_r655769753 but couldn't get diagnostic to show up for that, so didn't add any special handling)